code/iaas/service-discovery/server/src/main/java/io/cattle/platform/servicediscovery/deployment/impl/DeploymentManagerImpl.java

```diff
@@ -268,12 +269,26 @@ public void doWithLockNoResult() {
 
     @Override
     public void activateGlobalServicesForHost(long accountId, long hostId) {
+
+        class DeferredServiceActivate implements Runnable {
+            Service service;
+
+            public DeferredServiceActivate(Service service) {
+                this.service = service;
+            }
+
+            @Override
+            public void run() {
+                activate(service);
+            }
+        }
+
         List<? extends Service> services = expMapDao.getActiveServices(accountId);
         for (Service service: services) {
             Map<String, String> serviceLabels = sdSvc.getServiceLabels(service);
             if (serviceLabels.containsKey(ServiceDiscoveryConstants.LABEL_SERVICE_GLOBAL) &&
                     allocatorSvc.hostSatisfiesHostAffinity(hostId, serviceLabels)) {
-                activate(service);
+                DeferredUtils.defer(new DeferredServiceActivate(service));
```

Comment:
As a side note, global in general should use this approach to by calling DeploymentManager.reconcileServicesFor(account).  This method was added by me to handle health check but I had planned that global could use it too.